### DISCUSSION TITLE
Voxel_grid clustering value issue on gpu

### DISCUSSION THREAD

**Discussion by ericyj-lee** [pyg-team/pytorch_geometric#8903] [General]
This is a snippet of my code using [voxel_grid clustering](https://pytorch-geometric.readthedocs.io/en/latest/generated/torch_geometric.nn.pool.voxel_grid.html) with max_pool_x, for simplicity, I just let the data run through train_loader in the model without training the data at all.
<img width="468" alt="001" src="https://github.com/pyg-team/pytorch_geometric/assets/75584302/5e392b03-c477-46ae-9211-d2267a11318d">
<img width="298" alt="002" src="https://github.com/pyg-team/pytorch_geometric/assets/75584302/c0867127-c9b6-4231-9156-bdf9b23892f8">

The data.pos[x,y,z] of my data has maximum value of [127,127,1.5]. I aimed to cluster the nodes only based on [x,y] and I have tried the size parameters in voxel_grid of [32,32], [32,32,0] and [32,32,a large value(ex:100000)], 
1.	First, want to check which form is correct since these 3 work good for me when running my code on “cpu”, (issue happen when run on “gpu”, will explain later) 
2.	I group my data using PyG Dataloader of batch_size=64, and ideally the cluster number of each batch is (127/32)*(127/32) = 16, and through batches should be 16 * 64(batch_size) = 1024 as shown. 

<img width="323" alt="003" src="https://github.com/pyg-team/pytorch_geometric/assets/75584302/d64e33a1-196f-4722-92d8-326d5d50fc53">

However, sometimes when I run my code on “cuda:0”, it shows the following results.
<img width="407" alt="004" src="https://github.com/pyg-team/pytorch_geometric/assets/75584302/3c3f17ea-83ed-4734-8f2f-17e2e5559a84">

I analyze this issue by printing out more detailed relations of batch, data.pos ,and cluster(1st column in the picture)

<img width="248" alt="005" src="https://github.com/pyg-team/pytorch_geometric/assets/75584302/0df0d908-db8c-4a4c-b835-676ab30765c4">
<img width="292" alt="006" src="https://github.com/pyg-team/pytorch_geometric/assets/75584302/2acbfafc-f448-4bb2-a0fd-bb7f8cac03b4">

It seems that when assigning clusters in different batches, it does not start correctly (correct one batch 0: cluster 0-15, batch1:16-31,…) though it has correct cluster numbers in the above case(batch1: 485301~485310 has 16 clusters). Moreover, sometimes I even got negative numbers for clusters. This issue only happens when I run my code on cuda:0, and it could run perfectly on cpu. The error I got is shown below. These results would disable the max_pool_x function(global_max_pool will be fine since it only gets one node from each batch) from working properly when having the size parameter and further applying it to the Fully-connected layer. Thanks!

<img width="468" alt="007" src="https://github.com/pyg-team/pytorch_geometric/assets/75584302/8c8b998e-5d4f-4fb5-875c-2155dc721388">

3. Using the same code as above, I have 5 voxel_grid clustering step in my model
<img width="338" alt="008" src="https://github.com/pyg-team/pytorch_geometric/assets/75584302/a98a2781-82fa-4805-bf71-0f78c1dea473">

Sometimes I got the results of decreased batch size through these 5 steps, 
<img width="107" alt="009" src="https://github.com/pyg-team/pytorch_geometric/assets/75584302/af600c0e-ba53-4e01-9583-d5561e9174ad">

This also never happens when run on cpu. Want to ask why would this happen since the batch size should not change after voxel_grid clustering. Thanks!

**Comment by rusty1s** [comment]
Thanks for the issue and sorry for my late reply. This is definitely unexpected. Is this reproducible for the same data across several runs? If yes, do you mind sharing the point cloud in order to reproduce?
